### PR TITLE
Fix audio that may disappear after sleep and 10.14 audio supports and  misc.

### PR DESCRIPTION
**Pull request by soto2080**
1.The ALCPlugFix is customized for our device and is fully tested.
2.Remove plugins in VoodooPS2 that conflicts with VoodooI2C.
3.Add a hyperlink to a script that enable HiDPI and generate profile for display. Works very well on my machine. It makes the display setting more native like a real MacBook pro that you don't need RDM to change resolutions.

### CODE REVIEW

**Comment by Zerui18** [comment]
Thanks, audio works for me on Mojave with this.

However, VoodooI2C Trackpad doesn't seem to work on my computer (it never has in fact), in that when VoodooPS2 Trackpad is disabled/removed the trackpad is not functional anymore. However, VoodooI2C's touchscreen and mouse seem to work perfectly. 🧐

**Comment by Zerui18** [comment]
@soto2080 thanks again but I’m stuck on resolving kernel panic on boot. After the previous post I wanted to switch back to VoodooPS2Trackpad so I copied over one I had (with the same version) into the VoodooPS2Controller’s plugin folder. Afterwards every time I try to reboot into macOS the kernel panics, almost immediately after booting begins. The verbose log does not expose any obvious errors that resemble anything I could find online. I’ve also tried removing the Trackpad kext using Windows but it still shows as loaded in the verbose log.
To sum up, I’m not 100% sure the Trackpad kext is causing the error but that was the only modification before the panics started. Even after removing it from SLE macOS is still somehow able to load it.

Might anyone have an idea on how to proceed from here?

**Comment by soto2080** [comment]
How you inject yours kexts, as for me, I inject most kext via Clover at boot time except rehabman's BT and audio kext. So that you can easily disable what kext you don't want to inject.
If you place FakeSMC under L/E or S/L/E, Clover will not inject anything.

**Comment by Zerui18** [comment]
All my kexts are located in SLE. When I boot from USB, would the foreign kexts in SLE and LE still be loaded?

**Comment by Zerui18** [comment]
@soto2080 I'm trying to work around the kernel panic by booting with only kexts installed on my bootable USB (I've always been able to boot into macOS with it), however, the kernel still goes into panic almost right after boot begins. Furthermore in the loadedkexts section of the panic log I can still see kexts which have been removed from both clover/kexts and S/L/E.
Is there a way to boot macOS without kext caches when even safe mode can't get me in? (I think something's abnormal in the kext cache)
Thanks

**Comment by soto2080** [comment]
try this to rebuild cache.
sudo chmod -Rf 755 /S*/L*/E*
sudo chown -Rf 0:0 /S*/L*/E*
sudo chmod -Rf 755 /L*/E*
sudo chown -Rf 0:0 /L*/E*
sudo rm -Rf /S*/L*/PrelinkedKernels/*
sudo rm -Rf /S*/L*/Caches/com.apple.kext.caches/*
sudo touch -f /S*/L*/E*
sudo touch -f /L*/E*
sudo kextcache -Boot -U /
or use Rehabman's clover fork that have the option to boot without cache.

**Comment by Zerui18** [comment]
Thanks, I was able to rebuild the kext cache , however there’s still that strange kernel panic on boot so I’ve decided to do a fresh reinstall... Hopefully it goes smoothly this time :)

**Comment by Zerui18** [comment]
@soto2080 Thanks, first time getting VoodooI2C to work!

**Comment by Zerui18** [comment]
@soto2080 I appears that after updating to VoodooI2C (native), kernel panic occurs every time after computer tries to wake from sleep (with power button light off). I suspect that this is the issue: https://github.com/alexandred/VoodooI2C/issues/62 . May I know where is the source code for the VoodooI2C (native) you are using?

**Comment by soto2080** [comment]
@Zerui18 You need a patched version for Mojave, [Here you go](https://github.com/alexandred/VoodooI2C/issues/70).

**Comment by Zerui18** [comment]
I've found the source code but thanks anyways :)
Regarding the issue above, the patch proposed by 'lichengzhang2005' works.

**Comment by nos1609** [comment]
@soto2080 What about AppleALC? Have you also patched it for the new layout-id?

**Comment by soto2080** [comment]
@nos1609  vit9696 fixed it long time ago.

**Comment by soto2080** [comment]
@the-darkvoid  I modify your script and hda patches for 10.14 support.
Hope you can merge my commits.

**Comment by soto2080** [comment]
> My Machine is Dell XPS 9360, i7-7550U, QHD+, on Mojave, with the-darkvoid:master, I loaded
> `$sudo ./XPS9360.sh --patch-hda`
> to installed AppleHDA_ALC256.kext and CodecCommander.kext to L/E, and AppleALC.kext also required in Clover/kext/Others,
> 
> So It works but with two issues:
> 
> 1. Microphone not work ( never work from 10.13.X to 10.14 )
> 2. Returned from sleep, sound still works, but headset jack not  work ( before sleep, it works).
> 
> So with your changes, I fixed the second issue by merging changes only about audio parts, includes below ( **against the-darkvoid:master** ), though a re-plug is required ( **you?** ).
> 
> * modified:  XPS9360.sh                    -- Added the exactly steps to install ALCPlugfix etc.
> * modified:  audio/ahhcd.plist           -- Layout b
> * deleted:    audio/layout1.plist          -- deleted this
> * Added:     audio/ALCPlugFix
> * Added:     audio/good.win.ALCPlugFix.plist
> * Added:     audio/hda-verb
> * Added:     audio/layoutb.plist          -- copied from above layout1.plist, with Layout changed to 'b'
> 
> NOTE: I used layout-id "b" instead other values as it was there.
> 
> btw: is anyone got microphone works for XPS 9360?

As for your first problem, My microphone works fine since HS to Mojave, your model may differs with our one with Intel 8gen platform.

**Comment by dell951** [comment]
> > My Machine is Dell XPS 9360, i7-7550U, QHD+, on Mojave, with the-darkvoid:master, I loaded
> > `$sudo ./XPS9360.sh --patch-hda`
> > to installed AppleHDA_ALC256.kext and CodecCommander.kext to L/E, and AppleALC.kext also required in Clover/kext/Others,
> > So It works but with two issues:
> > 
> > 1. Microphone not work ( never work from 10.13.X to 10.14 )
> > 2. Returned from sleep, sound still works, but headset jack not  work ( before sleep, it works).
> > 
> > So with your changes, I fixed the second issue by merging changes only about audio parts, includes below ( **against the-darkvoid:master** ), though a re-plug is required ( **you?** ).
> > 
> > * modified:  XPS9360.sh                    -- Added the exactly steps to install ALCPlugfix etc.
> > * modified:  audio/ahhcd.plist           -- Layout b
> > * deleted:    audio/layout1.plist          -- deleted this
> > * Added:     audio/ALCPlugFix
> > * Added:     audio/good.win.ALCPlugFix.plist
> > * Added:     audio/hda-verb
> > * Added:     audio/layoutb.plist          -- copied from above layout1.plist, with Layout changed to 'b'
> > 
> > NOTE: I used layout-id "b" instead other values as it was there.
> > btw: is anyone got microphone works for XPS 9360?
> 
> As for your first problem, My microphone works fine since HS to Mojave, your model may differs with our one with Intel 8gen platform.
> And the second problem is that the fix had restriction that, your audio device must persistent between starting sleep and resuming from it.
> That is when you never (un)plug headphone when sleeping.

Thanks for the quick response, I just figured the microphone issue -- it might be a hardware issue as it's not working in Windows 10 as well,  might need to call Dell about it.

**Comment by soto2080** [comment]
@the-darkvoid  I dropped the change to VoodooPS2.
As for HiDPI, It can make it looks like 1440x810, just scaling but with 1920x1080.

ALCPLugFix is just a daemon that execute hda-verb, I don't know why CodecCommander not work sometime, but the one from goodwin works great.

And the last change is modify patchhda script to use layout id 2, that will work fine on every macOS.

**Comment by the-darkvoid** [comment]
@soto2080, thanks for working on this. I've cherry picked and merged the two commits which were fine.

**Comment by the-darkvoid** [comment]
Main repository now is integrated with ComboJack (development continuation of ALCPlugFix specifically for XPS 13).

Thanks @soto2080 for the idea of user-land daemon to handle headset events.
